sources/academy/webscraping/web_scraping_for_beginners/crawling/headless_browser.md

```diff
@@ -18,140 +18,180 @@ A headless browser is simply a browser that runs without a user interface (UI).
 
 ## Running in headless mode {#running-headless}
 
-We said that headless browsers didn't have a UI, but while scraping with the above scraper code, we're sure your screen was full of browser tabs. That's because in Crawlee, browsers run **headful** (with a UI) by default. This is useful for debugging and seeing what's going on in the browser. Once the scraper is complete, we switch it to **headless** mode using one of two methods:
-
-### Environment variable {#headless-env-var}
+We said that headless browsers didn't have a UI, but while scraping with the above scraper code, you could definitely see the browser. That's because we added the `headless: false` option. This is useful for debugging and seeing what's going on in the browser. Once your scraper is complete, you can remove the line and the crawler will run without a UI.
 
-This is the programmer's preferred solution, because you don't have to change the source code to change the behavior. All you need to do is set the [`CRAWLEE_HEADLESS=1`](https://crawlee.dev/docs/guides/configuration#crawlee_headless) environment variable when running your scraper, and it will automatically run headless. For example like this:
+You can also switch between headless and headful (with UI) using the [`CRAWLEE_HEADLESS`](https://crawlee.dev/docs/guides/configuration#crawlee_headless) environment variable. This allows you to change the mode without touching your code.
 
 <Tabs groupId="main">
 <TabItem value="MacOS/Linux" label="MacOS/Linux">
 
 ```shell
-CRAWLEE_HEADLESS=1 node crawlee.js
-
+CRAWLEE_HEADLESS=1 node browser.js
 ```
 
 </TabItem>
 <TabItem value="Windows CMD" label="Windows CMD">
 
 ```shell
-set CRAWLEE_HEADLESS=1 && node crawlee.js
-
+set CRAWLEE_HEADLESS=1 && node browser.js
 ```
 
 </TabItem>
 <TabItem value="Windows Powershell" label="Windows Powershell">
 
 ```shell
-$env:CRAWLEE_HEADLESS=1; & node crawlee.js
-
+$env:CRAWLEE_HEADLESS=1; & node browser.js
 ```
 
 </TabItem>
 </Tabs>
 
-### Scraper code {#headless-code}
+## Dynamically loaded data {#dynamic-data}
+
+One of the important benefits of using a browser is that it allows you to easily extract data that's dynamically loaded, such as data that's only fetched after a user scrolls or interacts with the page. In our case, it's the "**You may also like**" section of the product detail pages. Those products aren't available in the initial HTML, but the browser loads them later using an API.
+
+![headless-dynamic-data.png](./images/headless-dynamic-data.png)
+
+:::tip
+
+We discuss dynamic data at length in the [How to scrape dynamic pages](../../../tutorials/node_js/dealing_with_dynamic_pages.md) tutorial, and we also have a special lesson dedicated to it in our [Puppeteer & Playwright course](../../puppeteer_playwright/page/waiting.md).
+
+:::
+
+If we added an appropriate selector to our original `CheerioCrawler` code, it would not extract the information, but browser automatically fetches and renders this extra data.
```

Comment:
maybe a note about JSDOM somewhere here would be nice, with a mention of `runScripts` option that enables this behavior too, although it tends to fail on some sites so it's rather experimental

Reply:
I'd prefer not to confuse the readers at this point. It's for beginners.